### PR TITLE
Automatic grading for bundle item submission

### PR DESCRIPTION
**Pull request by nathanchrs**
Implements API endpoints for #112

- [x] Automatic grading of bundle item submission (currently supports statement and multiple choice item types)
- [x] Change bundle submission endpoint response format to be grouped by problem

## New/modified endpoints
- `/api/v2/contests/submissions/bundle`: for listing latest submissions with pagination
- `/api/v2/contests/submissions/bundle/summary`: for viewing contestant's latest answers for the current contest, grouped by problem
- `/api/v2/contests/submissions/bundle/answers`: for viewing contestant's latest answers for the current problem, grouped by item

### CODE REVIEW

File: judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/SetVerdictItemSubmissionGrader.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package judgels.sandalphon.submission.bundle;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.Optional;
+import judgels.sandalphon.api.problem.bundle.Item;
+import judgels.sandalphon.api.submission.bundle.ItemSubmission.Grading;
+import judgels.sandalphon.api.submission.bundle.Verdict;
+
+public class SetVerdictItemSubmissionGrader implements ItemSubmissionGrader {
```
Comment (fushar) [inline judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/SetVerdictItemSubmissionGrader.java]:
```suggestion
public class FixedVerdictItemSubmissionGrader implements ItemSubmissionGrader {
```

File: judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                .config("Invalid JSON")
+                .build();
+    }
+
+    @Test
+    void accepted() {
+        MultipleChoiceItemSubmissionGrader grader = new MultipleChoiceItemSubmissionGrader(mapper);
+
+        Grading grading = grader.grade(testItem1, "b");
+        assertThat(grading.getVerdict().getCode()).isEqualTo(Verdicts.ACCEPTED.getCode());
```
Comment (fushar) [inline judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java]:
```suggestion
        assertThat(grading.getVerdict()).isEqualTo(Verdicts.ACCEPTED);
```

(and all similar occurrences)

File: judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        assertThat(grading.getVerdict().getName()).isEqualTo(Verdicts.ACCEPTED.getName());
+        assertThat(grading.getScore().isPresent()).isTrue();
```
Comment (fushar) [inline judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java]:
```suggestion
        assertThat(grading.getScore()).contains(4);
```

(and all similar occurrences)

File: judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/ItemSubmission.java [hunk 4]
```diff
@@ -15,6 +16,20 @@
...
     Instant getTime();
+    Optional<Grading> getGrading();
+
+    static ItemSubmission withoutGrading(ItemSubmission submission) {
+        return new ItemSubmission.Builder().from(submission).grading(Optional.empty()).build();
+    }
+
+    @Value.Immutable
+    @JsonDeserialize(as = ImmutableGrading.class)
+    interface Grading {
```
Comment (fushar) [inline judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/ItemSubmission.java]: Why not make this as a top-level interface `ItemGrading`?

File: judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/Verdict.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package judgels.sandalphon.api.submission.bundle;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@JsonDeserialize(as = ImmutableVerdict.class)
+public interface Verdict {
+    String getCode();
+    String getName();
```
Comment (fushar) [inline judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/Verdict.java]: This is legacy way from programming. Actually we don't have to store the name in DB as it is actually a derived value. Frontend should provide the name is required. So please remove the name.

File: judgels-backends/uriel/uriel-app/src/main/resources/migrations/003_add-grading-to-bundle-item-submission.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+    <changeSet author="nathanchrs" id="add-grading-to-bundle-item-submission">
+        <addColumn catalogName="judgels_uriel" schemaName="judgels_uriel" tableName="uriel_contest_bundle_item_submission">
+            <column name="verdictCode" type="VARCHAR(255)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="verdictName" type="VARCHAR(255)">
```
Comment (fushar) [inline judgels-backends/uriel/uriel-app/src/main/resources/migrations/003_add-grading-to-bundle-item-submission.xml]: As explained before, this is not needed

**Review by fushar — CHANGES_REQUESTED** [review]
I have some questions about the submission endpoints; let's talk tomorrow

File: judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/SetVerdictItemSubmissionGrader.java (hunk 1 above)
Comment (nathanchrs) [reply judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/SetVerdictItemSubmissionGrader.java]: OK, changed

File: judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java (hunk 2 above)
Comment (nathanchrs) [reply judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java]: OK, fixed

File: judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java (hunk 3 above)
Comment (nathanchrs) [reply judgels-backends/sandalphon/sandalphon-client/src/test/java/judgels/sandalphon/submission/bundle/MultipleChoiceItemSubmissionGraderTests.java]: OK, fixed - turns out I was importing an older version of `assertThat`

File: judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/ItemSubmission.java (hunk 4 above)
Comment (nathanchrs) [reply judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/ItemSubmission.java]: OK, changed

File: judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/Verdict.java (hunk 5 above)
Comment (nathanchrs) [reply judgels-backends/sandalphon/sandalphon-api/src/main/java/judgels/sandalphon/api/submission/bundle/Verdict.java]: OK - changed Verdict to enum

**Review by fushar — APPROVED** [review]

**Pull request merged**